File: mac/Tests/BG3AssistantTests/RunCreationTests.swift
import XCTest
@testable import BG3HonorAssistant

final class RunCreationTests: XCTestCase {
    private let latestScores = AbilityScores(
        strength: 8, dexterity: 16, constitution: 14,
        intelligence: 10, wisdom: 10, charisma: 17
    )

    private func latestBuild() -> BuildSummary {
        BuildSummary(
            id: "latest-build",
            name: "Latest Swords Bard",
            honorStatus: "Reviewed",
            role: "Control",
            finalSplit: "Bard 12",
            classProgression: "Bard 1-12",
            startingAbilities: "",
            startingAbilityScores: latestScores,
            playPattern: "",
            caveat: "",
            source: "",
            levels: [
                BuildLevel(
                    level: 1, take: "Bard 1", subclassChoice: "-",
                    choices: "", tactics: "", confidence: "Reviewed"
                )
            ],
            gear: []
        )
    }

    func testFreshRunKeepsCharacterAndLatestBuildPresetsOnly() throws {
        var source = HonorRun()
        source.normalizeRoster()
        source.selectedAct = 3
        source.storyOutcomes = ["saved-grove"]
        source.walkthroughProgress = ["step": .completed]
        source.equippedByMember = ["tav": ["old-item"]]
        source.gearTarget = GearTarget(memberId: "tav", buildId: "latest-build", gearId: "old-item")

        let tavIndex = try XCTUnwrap(source.roster?.firstIndex { $0.id == "tav" })
        source.roster?[tavIndex].name = "Ariadne"
        source.roster?[tavIndex].level = 12
        source.roster?[tavIndex].buildId = "latest-build"
        source.roster?[tavIndex].className = "Old class"
        source.roster?[tavIndex].abilityScores = .customDefault
        source.roster?[tavIndex].abilityModifiers = [
            AbilityModifier(
                ability: .strength, kind: .permanent, mode: .add,
                value: 1, source: "Old reward"
            )
        ]
        source.roster?[tavIndex].appliedAbilitySetupId = "old-setup"

        let shadowheartIndex = try XCTUnwrap(source.roster?.firstIndex { $0.name == "Shadowheart" })
        source.roster?[shadowheartIndex].buildId = "removed-build"
        source.roster?[shadowheartIndex].status = .dead
        source.syncActivePartyProjection()

        let createdAt = Date(timeIntervalSince1970: 1234)
        let fresh = source.freshRun(
            name: "New Run",
            guideVersion: "latest-guide",
            availableBuilds: [latestBuild()],
            createdAt: createdAt
        )

        XCTAssertTrue(fresh.progress.isEmpty)
        XCTAssertTrue(fresh.walkthroughProgress?.isEmpty ?? true)
        XCTAssertTrue(fresh.storyOutcomes?.isEmpty ?? true)
        XCTAssertTrue(fresh.equippedByMember?.isEmpty ?? true)
        XCTAssertNil(fresh.gearTarget)

        let tav = try XCTUnwrap(fresh.roster?.first { $0.id == "tav" })
        XCTAssertEqual(tav.name, "Ariadne")
        XCTAssertEqual(tav.level, 1)
        XCTAssertEqual(tav.buildId, "latest-build")
        XCTAssertEqual(tav.className, "Bard 1")
        XCTAssertEqual(tav.abilityScores, latestScores)
        XCTAssertTrue(tav.abilityModifiers?.isEmpty ?? false)
        XCTAssertNil(tav.appliedAbilitySetupId)
        XCTAssertEqual(fresh.buildAssignedAt?[tav.id], createdAt)

        let shadowheart = try XCTUnwrap(fresh.roster?.first { $0.name == "Shadowheart" })
        XCTAssertNil(shadowheart.buildId)
        XCTAssertEqual(shadowheart.rosterStatus, .active)
    }
}

File: windows/tests/BG3HonorAssistant.Core.Tests/Models/RunCreationTests.cs
using BG3HonorAssistant.Core.Models;

namespace BG3HonorAssistant.Core.Tests.Models;

public sealed class RunCreationTests
{
    private static readonly AbilityScores LatestScores =
        new(8, 16, 14, 10, 10, 17);

    [Fact]
    public void FreshRunKeepsCharacterAndLatestBuildPresetsOnly()
    {
        var source = new HonorRun();
        source.NormalizeRoster();
        source.SelectedAct = 3;
        source.StoryOutcomes = ["saved-grove"];
        source.WalkthroughProgress = new Dictionary<string, CheckpointDisposition>
        {
            ["step"] = CheckpointDisposition.Completed,
        };
        source.EquippedByMember = new Dictionary<string, HashSet<string>>
        {
            ["tav"] = ["old-item"],
        };
        source.GearTarget = new GearTarget("tav", "latest-build", "old-item");

        var tavIndex = source.Roster!.FindIndex(member => member.Id == "tav");
        source.Roster[tavIndex] = source.Roster[tavIndex] with
        {
            Name = "Ariadne",
            Level = 12,
            BuildId = "latest-build",
            ClassName = "Old class",
            AbilityScores = AbilityScores.CustomDefault,
            AbilityModifiers =
            [
                new AbilityModifier(
                    Ability.Strength,
                    AbilityModifierKind.Permanent,
                    AbilityModifierMode.Add,
                    1,
                    "Old reward"),
            ],
            AppliedAbilitySetupId = "old-setup",
        };

        var shadowheartIndex = source.Roster.FindIndex(member => member.Name == "Shadowheart");
        source.Roster[shadowheartIndex] = source.Roster[shadowheartIndex] with
        {
            BuildId = "removed-build",
            Status = RosterStatus.Dead,
        };
        source.SyncActivePartyProjection();

        var createdAt = DateTimeOffset.FromUnixTimeSeconds(1234);
        var fresh = source.FreshRun(
            "New Run",
            "latest-guide",
            [LatestBuild()],
            createdAt);

        Assert.Empty(fresh.Progress);
        Assert.True(fresh.WalkthroughProgress is null or { Count: 0 });
        Assert.True(fresh.StoryOutcomes is null or { Count: 0 });
        Assert.True(fresh.EquippedByMember is null or { Count: 0 });
        Assert.Null(fresh.GearTarget);

        var tav = Assert.Single(fresh.Roster!, member => member.Id == "tav");
        Assert.Equal("Ariadne", tav.Name);
        Assert.Equal(1, tav.Level);
        Assert.Equal("latest-build", tav.BuildId);
        Assert.Equal("Bard 1", tav.ClassName);
        Assert.Equal(LatestScores, tav.AbilityScores);
        Assert.Empty(tav.AbilityModifiers!);
        Assert.Null(tav.AppliedAbilitySetupId);
        Assert.Equal(createdAt, fresh.BuildAssignedAt![tav.Id]);

        var shadowheart = Assert.Single(
            fresh.Roster!,
            member => member.Name == "Shadowheart");
        Assert.Null(shadowheart.BuildId);
        Assert.Equal(RosterStatus.Active, shadowheart.RosterStatus);
    }

    private static BuildSummary LatestBuild() =>
        new()
        {
            Id = "latest-build",
            Name = "Latest Swords Bard",
            HonorStatus = "Reviewed",
            Role = "Control",
            FinalSplit = "Bard 12",
            ClassProgression = "Bard 1-12",
            StartingAbilityScores = LatestScores,
            Levels =
            [
                new BuildLevel(1, "Bard 1", "-", string.Empty, string.Empty, "Reviewed"),
            ],
        };
}
